libs/vkd3d/va_map.c

```diff
@@ -0,0 +1,382 @@
+/*
+ * Copyright 2020 Philip Rebohle for Valve Software
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
+ */
+
+#define VKD3D_DBG_CHANNEL VKD3D_DBG_CHANNEL_API
+
+#include "vkd3d_private.h"
+
+static inline VkDeviceAddress vkd3d_va_map_get_next_address(VkDeviceAddress va)
+{
+    return va >> (VKD3D_VA_BLOCK_SIZE_BITS + VKD3D_VA_BLOCK_BITS);
+}
+
+static inline VkDeviceAddress vkd3d_va_map_get_block_address(VkDeviceAddress va)
+{
+    return (va >> VKD3D_VA_BLOCK_SIZE_BITS) & VKD3D_VA_BLOCK_MASK;
+}
+
+static const struct vkd3d_va_block *vkd3d_va_map_find_block(const struct vkd3d_va_map *va_map, VkDeviceAddress va)
+{
+    VkDeviceAddress next_address = vkd3d_va_map_get_next_address(va);
+    const struct vkd3d_va_tree *tree = &va_map->va_tree;
+
+    while (next_address && tree)
```

Comment:
Is the general idea here is that this will usually only take one iteration through the loop since this covers any VA up to 40 bits in the first iteration?

Reply:
Yes.